case_studies/dc2_cataloging/utils/variational_dist.py

```diff
@@ -1,232 +1,117 @@
+        nll_gating = rearrange(nll_gating, "b nth ntw -> b nth ntw 1 1")
+        return torch.where(nll_gating, vsbc, torch.nan)
 
-    @classmethod
-    def _fast_one_to_one_match(
-        cls, dist_matrix: torch.Tensor, est_is_on_mask: torch.Tensor, true_is_on_mask: torch.Tensor
-    ):
-        assert dist_matrix.shape[1] == dist_matrix.shape[2]
-        m = dist_matrix.shape[1]
-        _, match1 = torch.min(dist_matrix, dim=-2)  # (B, m)
-        _, match2 = torch.min(dist_matrix, dim=-1)  # (B, m)
-        matches = (match1, match2)
-        est_to_true_indices = torch.stack(matches, dim=-1)  # (B, m, 2) 2 represents index
-
-        # find one-to-one matching
-        match1cond2 = torch.gather(match1, 1, match2)  # (B, m)
-        pos_tensor = torch.arange(
-            0, m, dtype=est_to_true_indices.dtype, device=est_to_true_indices.device
-        ).view(
-            1, -1
-        )  # (1, m)
-        one_to_one_match = (match1cond2 == pos_tensor).unsqueeze(-1)  # (B, m, 1)
-        est_on_mask_num = rearrange(est_is_on_mask, "b nth ntw m 1 -> (b nth ntw) m").sum(
-            dim=-1, keepdim=True
-        )  # (B, 1)
-        true_on_mask_num = rearrange(true_is_on_mask, "b nth ntw m 1 -> (b nth ntw) m").sum(
-            dim=-1, keepdim=True
-        )
-        est_out_mask = est_to_true_indices[..., 0] > (est_on_mask_num - 1)  # (B, m)
-        true_out_mask = est_to_true_indices[..., 1] > (true_on_mask_num - 1)  # (B, m)
-        one_to_one_match &= ~(est_out_mask | true_out_mask).unsqueeze(-1)
-
-        offset = torch.where(one_to_one_match, 0, m)  # (B, m, 1)
-        return est_to_true_indices * one_to_one_match + offset
-
-    @classmethod
-    def _linear_sum_assignment(cls, dist_matrix: torch.Tensor):
-        ori_device = dist_matrix.device
-        dist_matrix_np = dist_matrix.cpu().numpy()
-        est_to_true_indices_list = []
-        for i in range(dist_matrix.shape[0]):
-            row_index, col_index = linear_sum_assignment(dist_matrix_np[i])
-            cur_indices = np.stack((row_index, col_index), axis=-1)
-            est_to_true_indices_list.append(cur_indices)
-        est_to_true_indices = np.stack(est_to_true_indices_list, axis=0)
-        return torch.from_numpy(est_to_true_indices).to(device=ori_device)
+    def sample_vsbc(self, x_cat: torch.Tensor, true_tile_cat: TileCatalog, use_mode=False):
+        assert true_tile_cat.max_sources == 1
+        sample_result = self.sample(x_cat, use_mode=use_mode)
+        assert sample_result.max_sources == 1
+        fp_pairs = self._factor_param_pairs(x_cat)
+        # for now, only test the vsbc for locs, ellipticity and flux
+        vsbc_variables = {
+            "locs",
+            "ellipticity",
+            "star_fluxes",
+            "galaxy_fluxes",
+        }
+        for qk, params in fp_pairs:
+            if qk.name in vsbc_variables:
+                sample_result[qk.name + "_vsbc"] = self._get_vsbc(qk, params, true_tile_cat)
```

Comment:
Does the VSBC metric deal with checkerboard structure of the variational distribution correctly? It seems tricky to apply VSBC with the checkerboard because we can't just read off the CDF of the marginal, as we could if the variational distribution were mean field

Reply:
I'll try to find a way to bypass it in the next PR.